### PR TITLE
Automatically produce a "complex" migration plan from simple ones

### PR DESCRIPTION
**Pull request by dralley**
closes: #6516
https://pulp.plan.io/issues/6516

### CODE REVIEW

**Comment by dralley** [comment]
The order things happen in is messed up, first I was having issues with accidental infinite recursion, now the opposite problem.

The code is conceptually finished though if you want to give it a brief look-over.

**Comment by dralley** [comment]
The problem right now is that the repositories are migrated but it doesn't create publications & distributions.  Hopefully that's not too difficult to resolve.

**Comment by dralley** [comment]
@goosemania @ipanova This is ready.  It's a little messy but structuring it more cleanly might be difficult due to the weird order everything needs to happen in.

File: pulp_2to3_migration/app/migration.py [hunk 1]
```diff
@@ -297,40 +260,36 @@ def create_repoversions_publications_distributions(plan, parallel=True):
...
+                    task_group=TaskGroup.current()
                 )
-                progress_dist.update(total=F('total') + dist_to_create)
-            else:
-                # Serial (non-parallel)
-                for repo_name in pulp3_repo_setup:
-                    task_args = [plugin, pulp3_repo_setup, repo_name]
-                    task_func(*task_args)
+        else:
+            # Serial (non-parallel)
```
Comment (ipanova) [inline pulp_2to3_migration/app/migration.py:281]: is there a reason to leave this piece of code? what would be the benefit of creating repos in serial way from complex plan?

Comment (dralley) [reply pulp_2to3_migration/app/migration.py:281]: It makes testing easier

**Comment by dralley** [comment]
Sidenote, how do we fix the redmine integration?  I need to manually change the issue states for issues in the migration tracker.

File: pulp_2to3_migration/app/migration.py [hunk 2]
```diff
@@ -158,42 +158,6 @@ def migrate_importers(plan):
                 pb.save()
 
 
-def simple_plugin_migration(plugin):
```
Comment (goosemania) [inline pulp_2to3_migration/app/migration.py:161]: :tada:

File: pulp_2to3_migration/app/pre_migration.py [hunk 3]
```diff
@@ -350,7 +348,7 @@ def pre_migrate_all_without_content(plan, type_to_repo_ids, repo_id_to_type):
         for plugin_plan in plan.get_plugin_plans():
             repos = plugin_plan.get_repositories()
             # filter by repo type
-            repos_to_check = type_to_repo_ids[plugin_plan.type]
+            repos_to_check = plan.type_to_repo_ids[plugin_plan.type]
```
Comment (goosemania) [inline pulp_2to3_migration/app/pre_migration.py:351]: :+1:

File: pulp_2to3_migration/app/models/base.py [hunk 4]
```diff
@@ -26,12 +83,54 @@ class MigrationPlan(BaseModel):
...
+                    name = repository['repo_id']
+                    plugin_plan.repositories_to_create[name] = {
+                        "pulp2_importer_repository_id": repository['repo_id'],
+                        "repository_versions": [
+                            {
+                                "repo_id": repository['repo_id'],
+                                "dist_repo_ids": [repository['repo_id']]
+                            }
+                        ]
+                    }
```
Comment (goosemania) [inline pulp_2to3_migration/app/models/base.py:114]: This makes sense to me.

**Review by goosemania — COMMENTED** [review]
@dralley, I haven't tested it but the code looks fine to me and makes sense.
I believe there are more places to cleanup the code paths which were specifically for simple plan but me can deal with it as we run into them.

Maybe it makes sense to rebase this PR (to run tests which I merged recently) and unskip the tests which are meant to cover the simple plan case?

**Comment by dralley** [comment]
@goosemania So the problem is that simple and complex plans still aren't quite identical (and they can't be).  Basically we can't determine the type of an empty repository, but if it's listed explicitly in the complex plan then it can be assumed.

I think what we need to do is keep the "content" part of the plan dictionary, and wrap it in an object that counts differently depending on the context.  Do you have any other ideas?

**Comment by goosemania** [comment]
Oh, you are right, there is still an issue with empty repos.
In theory, you can determine the type by its importer or distributor but I think it's not worth it, since empty repos are needed for Katello only and they always specify the complex plan.

> I think what we need to do is keep the "content" part of the plan dictionary, and wrap it in an object that counts differently depending on the context. Do you have any other ideas?

No.

**Comment by ipanova** [comment]
> Oh, you are right, there is still an issue with empty repos.
> In theory, you can determine the type by its importer or distributor but I think it's not worth it, since empty repos are needed for Katello only and they always specify the complex plan.
> 
> > I think what we need to do is keep the "content" part of the plan dictionary, and wrap it in an object that counts differently depending on the context. Do you have any other ideas?
> 
> No.

I was wondering if that would help to document the empty repo migration logic, somewhere along the lines:

"Empty repos won't be migrated by default unless specifically mentioned in the plan"

**Comment by ipanova** [comment]
I have locally tested and the conversion works as expected. Just empty repos are not migrated unless directly specified in the plan.

Not sure whether this will be a benefit or maybe overkill, but maybe we should return the converted plan?
```
$ http POST :24817/pulp/api/v3/migration-plans/ plan='{"plugins": [{"type": "rpm"}]}'
HTTP/1.1 201 Created
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, POST, HEAD, OPTIONS
Connection: close
Content-Length: 163
Content-Type: application/json
Correlation-ID: 200e445d396a4d028880b99992d88d7d
Date: Thu, 14 Jan 2021 13:44:19 GMT
Location: /pulp/api/v3/migration-plans/30e56a5a-c1f4-44c1-87cf-0407374a041a/
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "plan": {
        "plugins": [
            {
                "type": "rpm"
            }
        ]
    },
    "pulp_created": "2021-01-14T13:44:19.780714Z",
    "pulp_href": "/pulp/api/v3/migration-plans/30e56a5a-c1f4-44c1-87cf-0407374a041a/"
}
```

**Comment by goosemania** [comment]
> I have locally tested and the conversion works as expected. Just empty repos are not migrated unless directly specified in the plan.
> 
> Not sure whether this will be a benefit or maybe overkill, but maybe we should return the converted plan?
> 
> ```
> $ http POST :24817/pulp/api/v3/migration-plans/ plan='{"plugins": [{"type": "rpm"}]}'
> HTTP/1.1 201 Created
> Access-Control-Expose-Headers: Correlation-ID
> Allow: GET, POST, HEAD, OPTIONS
> Connection: close
> Content-Length: 163
> Content-Type: application/json
> Correlation-ID: 200e445d396a4d028880b99992d88d7d
> Date: Thu, 14 Jan 2021 13:44:19 GMT
> Location: /pulp/api/v3/migration-plans/30e56a5a-c1f4-44c1-87cf-0407374a041a/
> Server: gunicorn/20.0.4
> Vary: Accept, Cookie
> X-Frame-Options: SAMEORIGIN
> 
> {
>     "plan": {
>         "plugins": [
>             {
>                 "type": "rpm"
>             }
>         ]
>     },
>     "pulp_created": "2021-01-14T13:44:19.780714Z",
>     "pulp_href": "/pulp/api/v3/migration-plans/30e56a5a-c1f4-44c1-87cf-0407374a041a/"
> }
> ```

I believe in the past we settled on not doing it and leaving it as an option for future RFE if any user will really need it.
One of the reasons is the large output which comes quite unexpectedly and they likely don't really need it.

**Comment by ipanova** [comment]
> > I have locally tested and the conversion works as expected. Just empty repos are not migrated unless directly specified in the plan.
> > Not sure whether this will be a benefit or maybe overkill, but maybe we should return the converted plan?
> > ```
> > $ http POST :24817/pulp/api/v3/migration-plans/ plan='{"plugins": [{"type": "rpm"}]}'
> > HTTP/1.1 201 Created
> > Access-Control-Expose-Headers: Correlation-ID
> > Allow: GET, POST, HEAD, OPTIONS
> > Connection: close
> > Content-Length: 163
> > Content-Type: application/json
> > Correlation-ID: 200e445d396a4d028880b99992d88d7d
> > Date: Thu, 14 Jan 2021 13:44:19 GMT
> > Location: /pulp/api/v3/migration-plans/30e56a5a-c1f4-44c1-87cf-0407374a041a/
> > Server: gunicorn/20.0.4
> > Vary: Accept, Cookie
> > X-Frame-Options: SAMEORIGIN
> > 
> > {
> >     "plan": {
> >         "plugins": [
> >             {
> >                 "type": "rpm"
> >             }
> >         ]
> >     },
> >     "pulp_created": "2021-01-14T13:44:19.780714Z",
> >     "pulp_href": "/pulp/api/v3/migration-plans/30e56a5a-c1f4-44c1-87cf-0407374a041a/"
> > }
> > ```
> 
> I believe in the past we settled on not doing it and leaving it as an option for future RFE if any user will really need it.
> One of the reasons is the large output which comes quite unexpectedly and they likely don't really need it.

yes, this reasoning totally makes sense to me.

**Comment by dralley** [comment]
Also because all that work has to happen in the space of the web request, and it would be quite a lot of queries.

**Comment by dralley** [comment]
@goosemania Does the new commit look about right?

Still working on the other set of tests since it's more complex -- but I think the general idea would work.

**Comment by goosemania** [comment]
> @goosemania Does the new commit look about right?
> 
> Still working on the other set of tests since it's more complex -- but I think the general idea would work.

Yeah, looks right to me. 
More complex than initially anticipated indeed. Probably still better then just duplicating all the data and tests.
Thanks for working on it!

**Comment by dralley** [comment]
Not entirely convinced all that work with the tests is worth it, over just having a separate set of expectations : /  For the rerun tests at least.  Those were far more frustrating than the others.

But it's done now.

**Comment by goosemania** [comment]
> Not entirely convinced all that work with the tests is worth it, over just having a separate set of expectations : / For the rerun tests at least. Those were far more frustrating than the others.

Agreed :(

**Review by goosemania — APPROVED** [review]
Thanks! :octocat:

**Pull request merged**